fvm/src/call_manager/default.rs

```diff
@@ -471,6 +477,153 @@ where
+
+            // Make a store.
+            let mut store = engine.new_store(kernel);
+
+            let from_actor = self.machine().state_tree().get_actor(&msg.from)??;
+
+            // From this point on, there are no more syscall errors, only aborts.
+            let result: std::result::Result<BlockId, Abort> = (|| {
+                // Instantiate the module.
+                let instance = engine
+                    .get_instance(&mut store, &from_actor.code)
+                    .and_then(|i| i.context("actor code not found"))
+                    .map_err(Abort::Fatal)?;
+
+                // TODO: what memory is being resolved here?
+                // Resolve and store a reference to the exported memory.
+                let memory = instance
+                    .get_memory(&mut store, "memory")
+                    .context("actor has no memory export")
+                    .map_err(Abort::Fatal)?;
+                store.data_mut().memory = memory;
+
+                // Lookup the invoke method.
+                let validate: wasmtime::TypedFunc<(u32, u32, u32, u32), (u32, u32)> = instance
+                    .get_typed_func(&mut store, "validate")
+                    // All actors will have an invoke method.
+                    .map_err(Abort::Fatal)?;
+
+                // Set the available gas.
+                update_gas_available(&mut store)?;
+
+                // Invoke it.
+                let res = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+                    validate.call(&mut store, (params_id,))
+                }))
+                .map_err(|panic| Abort::Fatal(anyhow!("panic within actor: {:?}", panic)))?;
+
+                // Charge for any remaining uncharged execution gas, returning an error if we run
+                // out.
+                charge_for_exec(&mut store)?;
+
+                // If the invocation failed due to running out of exec_units, we have already
+                // detected it and returned OutOfGas above. Any other invocation failure is returned
+                // here as an Abort
+                Ok(res?)
+            })();
+
+            let invocation_data = store.into_data();
+            let last_error = invocation_data.last_error;
+            let (mut cm, block_registry) = invocation_data.kernel.into_inner();
+
+            // Resolve the return block's ID into an actual block, converting to an abort if it
+            // doesn't exist.
+            let result = result.and_then(|ret_id| {
+                Ok(if ret_id == NO_DATA_BLOCK_ID {
+                    None
+                } else {
+                    Some(block_registry.get(ret_id).map_err(|_| {
+                        Abort::Exit(
+                            ExitCode::SYS_MISSING_RETURN,
+                            String::from("returned block does not exist"),
+                        )
+                    })?)
```

Comment:
Could you create an issue as a holding pen for the error conditions we need to think through and model for?